### PR TITLE
[WIP] - Restructuring

### PR DESCRIPTION
**Pull request by jcrist**
Restructuring to be more maintainable.

### CODE REVIEW

File: dklearn/grid_search.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    X_train_name = 'X_train-' + token
+    y_train_name = 'y_train-' + token
+    X_test_name = 'X_test-' + token
+    y_test_name = 'y_test-' + token
+
+    train, test, dsk = unpack_arguments(train, test)
+    dsk[X_train_name] = (safe_indexing, X_name, train)
+    dsk[y_train_name] = (safe_indexing, y_name, train)
+    dsk[X_test_name] = (safe_indexing, X_name, test)
+    dsk[y_test_name] = (safe_indexing, y_name, test)
```
Comment (mrocklin) [inline dklearn/grid_search.py]: Out of curiosity, what ends up stopping you from using dask.delayed everywhere?  Is it naming?  Efficiency in handling graphs?

Comment (mrocklin) [inline dklearn/grid_search.py]: Or, what would need to be done to dask.delayed to make it a sufficient tool for this?

**Comment by mrocklin** [comment]
Things here seem pretty clean to me.  I've made some small questions but generally I think we should merge this sooner rather than later.

What is left in branches?  Logistic regression things?

File: dklearn/grid_search.py (hunk 1 above)
Comment (jcrist) [inline dklearn/grid_search.py]: A mix of both of those. Biggest one is avoiding repeating `tokenize` calls on parameters. These can add up.

Comment (mrocklin) [inline dklearn/grid_search.py]: That's surprising.  Parameters meaning the scalar params or parameters meaning numpy arrays?

**Comment by jcrist** [comment]
> What is left in branches? Logistic regression things?

Logistic regression, "meta-estimators" like `Averaged` and `Chained`, some work on a new `Matrix` class, and a branch that works more like `xarray`, with mutation and dask behind the scenes.

I plan to merge this now, and may pull in a few of the other things over the next couple days.

**Comment by mrocklin** [comment]
Hooray for merging

File: dklearn/grid_search.py (hunk 1 above)
Comment (jcrist) [inline dklearn/grid_search.py]: Numpy arrays - large numpy arrays can take a few ms to over a second. This is easy to work around though, by using `X = delayed(X)`, and then passing the delayed object around instead. I found the method used here more readable (and clear what was happening), but either could be used.

Comment (mrocklin) [inline dklearn/grid_search.py]: I see two arguments for delayed:

1.  It makes it easier for non-dask developers to jump in (which I think we should make a general goal of this project; we don't want to maintain this solo for forever)
2.  It forces us to improve the experience of writing delayed-backed libraries

**Pull request merged**